### PR TITLE
Comments GUI in files sidebar

### PR DESCRIPTION
**Pull request by PVince81**
Preliminary work for bringing comments in the files sidebar
- [x] TODO: pagination with REPORT method for pagination (currently uses PROPFIND)
- [x] TODO: styling of input field + comments row
- [x] TODO: unit tests

@blizzz

### CODE REVIEW

**Comment by PVince81** [comment]
Currently looks shitty
comments-v0

**Comment by karlitschek** [comment]
@jancborchardt What do you think?

**Comment by PVince81** [comment]
- [ ] TODO: allow owner to edit/delete their comments ?

**Comment by PVince81** [comment]
- [ ] TODO: mark as read logic

**Comment by zander** [comment]
Can someone remove me from @mention-bot, I have no idea why it thinks I have any input on this git repo.

File: core/shipped.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
     "activity",
     "admin_audit",
     "encryption",
+	"comments",
```
Comment (LukasReschke) [inline core/shipped.json]: Indentation

**Comment by LukasReschke** [comment]
> Can someone remove me from @mention-bot, I have no idea why it thinks I have any input on this git repo.

The bot looks who has commited to files. GitHub claims you have comitted to `.gitignore` as can be seen at https://github.com/owncloud/core/commits/master?author=zander.

To remove yourself add yourself at https://github.com/owncloud/core/blob/816c23c17a71443d8774107d54ada8bda7486f32/.mention-bot#L19-L21 and create a Pull Request.

File: apps/comments/js/commentstabview.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+			} else {
+				this.render();
+				this.collection.reset();
+			}
+		},
+
+		render: function() {
+			this.$el.html(this.template({
+				emptyResultLabel: t('comments', 'No other comments available'),
+				moreLabel: t('comments', 'More comments...')
```
Comment (LukasReschke) [inline apps/comments/js/commentstabview.js]: `…` instead of `...`

File: apps/comments/js/commentstabview.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		},
+
+		_onRequest: function() {
+			this._toggleLoading(true);
+			this.$el.find('.showMore').addClass('hidden');
+		},
+
+		_onEndRequest: function() {
+			this._toggleLoading(false);
+			this.$el.find('.empty').toggleClass('hidden', !!this.collection.length);
```
Comment (LukasReschke) [inline apps/comments/js/commentstabview.js]: JS magic or typo? :see_no_evil:

File: apps/comments/js/commentstabview.js (hunk 2 above)
Comment (PVince81) [inline apps/comments/js/commentstabview.js]: @LukasReschke you're too early to the party, this is very, very rough.
Also this was copy-pasted from the versions tab :wink:

File: apps/comments/js/commentstabview.js (hunk 3 above)
Comment (PVince81) [inline apps/comments/js/commentstabview.js]: `!!` is JS magic that coerces a value to become a boolean

**Comment by PVince81** [comment]
@blizzz it looks like the backend doesn't work properly with pagination.
To test: post a series of comments "One", "Two", "Three" up to "Ten".
Refresh the page: "Ten" is at the top as expected.
Then click "More comments".
The JS code will send a REPORT call with `<oc:datetime>` using the timestamp of the last shown comment, "Six".

Expected result: Server must return the next page containing "Five", "Four", etc.
Current result: Server returns "Ten", "Nine" again.

**Comment by PVince81** [comment]
Ref: https://github.com/owncloud/core/issues/16328

**Comment by PVince81** [comment]
- [ ] add avatars
- [ ] deleted user case

**Comment by PVince81** [comment]
- [ ] comment limits as per the requirements in https://github.com/owncloud/core/issues/16328#issue-76013209: max 1000 chars, only 250 displayed
- [ ] popup approach ? see https://github.com/owncloud/core/issues/16328#issue-76013209

**Comment by PVince81** [comment]
- [ ] REQUIRES: https://github.com/evert/davclient.js/pull/16

**Comment by jancborchardt** [comment]
I mean … I still think it’s bad that we have a separate view for Activity and Comments. The layout is _exactly_ the same. And what do we show as the default view? If we continue to show Activity, then new comments will be less visible, and the whole feature less findable.

This results in two different features battling for attention and people unnecessarily needing to switch tabs. I mentioned that in the past as well – hope that the plan is this being a stop-gap, and having a proper combined »Activity« view for 9.0.

Sure, I can just apply the same layout here by using the same classes as the activity view, but this is a larger interaction problem and just not polished. We keep adding more and more things …

**Comment by blizzz** [comment]
> @blizzz it looks like the backend doesn't work properly with pagination.
To test: post a series of comments "One", "Two", "Three" up to "Ten".
Refresh the page: "Ten" is at the top as expected.
Then click "More comments".
The JS code will send a REPORT call with <oc:datetime> using the timestamp of the last shown comment, "Six".

> Expected result: Server must return the next page containing "Five", "Four", etc.
Current result: Server returns "Ten", "Nine" again.

That's a misunderstand of what happens. The <oc:datetime> settings means "not older than". 

Because you look from the most current one and want to go down the tree, you need to adjust the offset instead. Example output with limit 4 and offset first 0, then 4:

```
curl -u master:master -i --data-binary "@report2.xml" -X REPORT -H "Content-Type: text/xml" http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/4639 2>/dev/null | grep message
    <oc:message>Ten</oc:message>
    <oc:message>Nine</oc:message>
    <oc:message>Eight</oc:message>
    <oc:message>Seven</oc:message>

curl -u master:master -i --data-binary "@report2.xml" -X REPORT -H "Content-Type: text/xml" http://zara.owncloud.bzoc/master/remote.php/dav/comments/files/4639 2>/dev/null | grep message
    <oc:message>Six</oc:message>
    <oc:message>Fife</oc:message>
    <oc:message>Four</oc:message>
    <oc:message>Three</oc:message>
```

**Comment by PVince81** [comment]
@blizzz I discussed pagination yesterday with @nickvergessen and it's usually better to base it on the mtime (or even the id) to have a more accurate offset.

That's why I thought that by specifying the "datetime" it would simply fetch the next 5 entries older than that date. @nickvergessen mind to comment ?

**Comment by PVince81** [comment]
@jancborchardt ok as a next step I'll apply the same layout like activities.
How to handle multi-line comments ? The original mock up only shows a single line for comments.

**Comment by blizzz** [comment]
> @blizzz I discussed pagination yesterday with @nickvergessen and it's usually better to base it on the mtime (or even the id) to have a more accurate offset.

> That's why I thought that by specifying the "datetime" it would simply fetch the next 5 entries older than that date. @nickvergessen mind to comment ?

Fetching items based on timestamps pulls in the problem of comments with timestamp clash. Some comments might not be loaded because of the same timestamp, but they did not do it into the result set because of limit.

If you requests comments starting the same last timestamp you pull in comments multiple times, uselessly. With the very unlikely situation of having more comments than the specified limit with the same creation time, fetching would be rendered broken as you pull the same stuff.

With custom IDs as per custom implementation in mind, you cannot rely on sequential IDs. You can exclude known IDs, but this would make the whole procedure far more complex.

Traditional limit-offset-combo does not bring this difficulties. As long as there is no real big advantage I'd prefer to keep this as is and go on with the remaining stuff.

**Comment by PVince81** [comment]
Now IIRC comments are supposed to also publish an activity. What we can do is show "Activity" by default, displaying shortened comments (single line). If a user sees a comment there and clicks on it (or three dots "...") it will then switch to the "Comments" tab and scroll down + highlight the selected comment.

In this case the "Activity" view would stay the default.

**Comment by blizzz** [comment]
> Now IIRC comments are supposed to also publish an activity. What we can do is show "Activity" by default, displaying shortened comments (single line). If a user sees a comment there and clicks on it (or three dots "...") it will then switch to the "Comments" tab and scroll down + highlight the selected comment.

Sounds interesting. Comments are still advertised, and Activity stays not-interactive as it is now (except you really want to change this). Avoids also bloat on that side, and specialized tab can take care of specific comment stuff. Per se I would say that "integrated" or interactive activity is great, for faster response etc., however if everything is mixed up there and potentially more items are getting inside, a clear look an a specific feature might be convenient.

**Comment by karlitschek** [comment]
I think this is a good idea. At least in the near term before we move to a combined view as @jancborchardt suggested. But this is something for a later release

**Comment by jancborchardt** [comment]
> What we can do is show "Activity" by default, displaying shortened comments (single line). If a user sees a comment there and clicks on it (or three dots "...") it will then switch to the "Comments" tab and scroll down + highlight the selected comment.

This sounds like it would be more difficult than actually just putting the Comments in there directly.

From a UX point of view it’s majorly unclean too, duplicating content. Why can I _view_ the comments in this tab, but need to switch to the other to view them fully, or write new ones?

As for displaying the comments, we should display max 3 lines, after that ellipsize. Clicking a comment will simply expand it. Nothing fancy here, and no separate views needed.

(I don’t mean to sound aggressive here … even though I kinda am. We talk about Slack / Dropbox and others all the time – then we also need to do it properly. Having this stuff in separate timelines just because it’s technically easier it not doing that. I also don’t think that this is something which should wait for the next release, especially when it has been specced integrated like that since some time.)

**Comment by PVince81** [comment]
We never really reached a full agreement on the timeline approach.

I'm still trying to gather pro/cons technically, here: https://github.com/owncloud/core/issues/20259#issuecomment-176721024.

In theory we could use the oc_activity table/view as the timeline view since we're going to push comments/tags/etc there. The problem with activities is that they expire, and comments might be detached from their answers if activity happens in-between.

**Comment by nickvergessen** [comment]
Just to make this sure: the activity app currently stores on activity per user, so commenting on activities (with activity, id as comment object) will not work, or at least each user just sees his own comments. Also it is not a timeline and needs a lot more changes before it can really be one...

So let's not over complicate this one week before the feature freeze... we can do activities that someone commented, we can also display the content of the comment as the long "message" of the activity (I'm doing that with announcements already as well). Switching to the comments-sidebar requires some magic in the activity sidebar, but I guess we can also do that.

**Comment by PVince81** [comment]
comments-betterlooking

Starts to look better. However I don't see how I could cram 3 rows within a single one.

- [ ] TODO: Ellipsize 3 rows somehow (might need height calculation logic), on click expand it

**Comment by PVince81** [comment]
Added unit tests + spinner when posting.

**Comment by PVince81** [comment]
This PR is ready for review, it's in a stable state.
Other enhancements will be processed separately (see checkboxes)

Please review @blizzz @icewind1991 @rullzer @nickvergessen

**Comment by icewind1991** [comment]
It fails to load comments for me since the root element in the request body is wrong (from what I can tell) and the check at https://github.com/owncloud/core/blob/24025f32f448d9bef1087bea01d771dd7e871996/apps/dav/lib/comments/commentsplugin.php#L152-L152 fails

**Comment by PVince81** [comment]
@icewind1991 works for me, maybe you switched branches and are in a strange state ?
It should work if you setup OC from scratch.

**Comment by PVince81** [comment]
After rebasing onto master, I get 415 unsupported media type.

Let me have a look.

**Comment by PVince81** [comment]
Rebased + fixed report root element name.

@icewind1991 please try again :smile:

**Comment by icewind1991** [comment]
Works now :+1:

**Comment by PVince81** [comment]
Move remaining items to https://github.com/owncloud/core/issues/16328#issuecomment-179120807

**Comment by PVince81** [comment]
Looks like I anticipated @blizzz's :+1: after talking with him on IRC, and was probably too tired and mistakenly thought he already reviewed it. :tired_face: 

@blizzz can you have a quick look when you have time ?

**Pull request merged**